x/mongo/driver/topology/pool.go

```diff
@@ -742,6 +742,57 @@ func (p *pool) removeConnection(conn *connection, reason reason, err error) erro
 	return nil
 }
 
+// BGCallback is a callback for monitoring the behavior of the experimental
+// background-read-on-timeout connection preserving mechanism.
+//
+// Deprecated: BGCallback is intended for experimental use only and may be
+// removed or modified at any time.
+var BGCallback func(addr string, start, read time.Time, errs []error, connClosed bool)
+
+// bgRead sets a new read deadline on the provided connection (1 second in the
+// future) and tries to read any bytes returned by the server. If successful, it
+// checks the connection into the provided pool. If there are any errors, close
+// the connection.
+func bgRead(pool *pool, conn *connection) {
+	var start, read time.Time
+	start = time.Now()
+	errs := make([]error, 0)
+	connClosed := false
+
+	err := conn.nc.SetReadDeadline(time.Now().Add(1 * time.Second))
+	if err != nil {
+		errs = append(errs, fmt.Errorf("error setting read deadline: %w", err))
+		err = conn.close()
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error closing after setting read deadline: %w", err))
+		}
+	}
+
+	if err == nil {
```

Comment:
The control flow comes here if it fails on L762 but passes on L765 when the `conn` is closed for a `SetReadDeadline` failure. The `conn` will try to read again on L775 underneath. It is not like the expected logic.

Reply:
Good catch! I will refactor that function to make the logic simpler.

Reply:
Fixed.